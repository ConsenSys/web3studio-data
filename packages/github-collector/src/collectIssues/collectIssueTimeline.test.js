const { toArray, take } = require('rxjs/operators');
const collectRepos = require('../collectRepos');
const collectIssues = require('./index');
const mockGithub = require('../../test/fixtures/mockGithub');
const repositoryIssues = require('../../test/fixtures/repositoryIssues');
const TestTransport = require('../../test/TestTransport');

describe('collectIssueTimeline', () => {
  let source;
  let transport;

  /**
   * Create a source Observable for the test suite
   *
   * @returns {Observable} Test Observable
   */
  const createSource = () =>
    collectRepos().pipe(
      take(1),
      collectIssues(transport),
      toArray()
    );

  afterAll(mockGithub.afterAll);
  beforeAll(mockGithub.beforeAll);

  beforeEach(() => {
    mockGithub.beforeEach();
    transport = new TestTransport();
    source = createSource();
  });

  it('Collects repo issue timelines and passes as events', async () => {
    const results = await source.toPromise();

    expect(results.length).toEqual(34);

    expect(results).toMatchSnapshot();
  });

  it('Picks up from the last cursor', async () => {
    transport.mockLastEvent = { cursor: 'someCursor' };

    await source.toPromise();

    const requests = mockGithub.requests();

    const hasCursor = requests.some(
      request => request.variables.cursor === transport.mockLastEvent.cursor
    );

    expect(hasCursor).toEqual(true);
  });

  it('skips timeline requests when the last cursor exists', async () => {
    const {
      data: {
        repository: {
          issues: { edges: issues }
        }
      }
    } = repositoryIssues();

    transport.mockLastEvent = {
      // Phew, that's a long path
      cursor: issues[0].node.timeline.edges[0].cursor
    };

    await source.toPromise();

    const requests = mockGithub.requests();

    // `issues.length`, total possible timeline fetches,
    // `- 1`, we're not fetching one b/c of the cursor
    // `+ 2` repo + issue fetch
    expect(requests.length).toEqual(issues.length - 1 + 2);
  });
});
