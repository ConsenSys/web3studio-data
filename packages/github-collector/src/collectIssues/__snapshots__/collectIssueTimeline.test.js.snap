// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collectIssueTimeline Collects repo issue timelines and passes as events 1`] = `
Array [
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "Mg",
      "issueId": "MDU6SXNzdWUzOTg0MjczNzY=",
      "issueNumber": 1,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-04T13:58:09Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "NA",
      "issueId": "MDU6SXNzdWUzOTg0MjczNzY=",
      "issueNumber": 1,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-18T17:20:16Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "Mg",
      "issueId": "MDU6SXNzdWUzOTkyNzY0ODQ=",
      "issueNumber": 3,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-04T13:58:09Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "NA",
      "issueId": "MDU6SXNzdWUzOTkyNzY0ODQ=",
      "issueNumber": 3,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-18T17:20:16Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "Mg",
      "issueId": "MDU6SXNzdWUzOTkyOTI4MjM=",
      "issueNumber": 4,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-04T13:58:09Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "NA",
      "issueId": "MDU6SXNzdWUzOTkyOTI4MjM=",
      "issueNumber": 4,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-18T17:20:16Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "Mg",
      "issueId": "MDU6SXNzdWUzOTk0ODExNDU=",
      "issueNumber": 5,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-04T13:58:09Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "NA",
      "issueId": "MDU6SXNzdWUzOTk0ODExNDU=",
      "issueNumber": 5,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-18T17:20:16Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "Mg",
      "issueId": "MDU6SXNzdWUzOTk1MjYwNjU=",
      "issueNumber": 8,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-04T13:58:09Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "NA",
      "issueId": "MDU6SXNzdWUzOTk1MjYwNjU=",
      "issueNumber": 8,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-18T17:20:16Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "Mg",
      "issueId": "MDU6SXNzdWUzOTk1MjYzMTU=",
      "issueNumber": 9,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-04T13:58:09Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "NA",
      "issueId": "MDU6SXNzdWUzOTk1MjYzMTU=",
      "issueNumber": 9,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-18T17:20:16Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "Mg",
      "issueId": "MDU6SXNzdWUzOTk1MjY2MzI=",
      "issueNumber": 10,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-04T13:58:09Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "NA",
      "issueId": "MDU6SXNzdWUzOTk1MjY2MzI=",
      "issueNumber": 10,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-18T17:20:16Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "Mg",
      "issueId": "MDU6SXNzdWUzOTk5MDczMzY=",
      "issueNumber": 11,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-04T13:58:09Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "NA",
      "issueId": "MDU6SXNzdWUzOTk5MDczMzY=",
      "issueNumber": 11,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-18T17:20:16Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "Mg",
      "issueId": "MDU6SXNzdWUzOTk5MTY1MTE=",
      "issueNumber": 12,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-04T13:58:09Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "NA",
      "issueId": "MDU6SXNzdWUzOTk5MTY1MTE=",
      "issueNumber": 12,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-18T17:20:16Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "Mg",
      "issueId": "MDU6SXNzdWUzOTk5NzQyMDM=",
      "issueNumber": 13,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-04T13:58:09Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "NA",
      "issueId": "MDU6SXNzdWUzOTk5NzQyMDM=",
      "issueNumber": 13,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-18T17:20:16Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "Mg",
      "issueId": "MDU6SXNzdWUzOTk5NzU3MjI=",
      "issueNumber": 14,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-04T13:58:09Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "NA",
      "issueId": "MDU6SXNzdWUzOTk5NzU3MjI=",
      "issueNumber": 14,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-18T17:20:16Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "Mg",
      "issueId": "MDU6SXNzdWUzOTk5ODE2MTA=",
      "issueNumber": 15,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-04T13:58:09Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "NA",
      "issueId": "MDU6SXNzdWUzOTk5ODE2MTA=",
      "issueNumber": 15,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-18T17:20:16Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "Mg",
      "issueId": "MDU6SXNzdWU0MDAzMzA5OTQ=",
      "issueNumber": 16,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-04T13:58:09Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "NA",
      "issueId": "MDU6SXNzdWU0MDAzMzA5OTQ=",
      "issueNumber": 16,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-18T17:20:16Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "Mg",
      "issueId": "MDU6SXNzdWU0MDA0NDA2NzY=",
      "issueNumber": 17,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-04T13:58:09Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "NA",
      "issueId": "MDU6SXNzdWU0MDA0NDA2NzY=",
      "issueNumber": 17,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-18T17:20:16Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "Mg",
      "issueId": "MDU6SXNzdWU0MDA0NDEwMTg=",
      "issueNumber": 18,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-04T13:58:09Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "NA",
      "issueId": "MDU6SXNzdWU0MDA0NDEwMTg=",
      "issueNumber": 18,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-18T17:20:16Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "Mg",
      "issueId": "MDU6SXNzdWU0MDA0NDY5NTE=",
      "issueNumber": 19,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-04T13:58:09Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "NA",
      "issueId": "MDU6SXNzdWU0MDA0NDY5NTE=",
      "issueNumber": 19,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-18T17:20:16Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "Mg",
      "issueId": "MDU6SXNzdWU0MDA1MTYyOTE=",
      "issueNumber": 20,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-04T13:58:09Z",
  },
  Object {
    "event": "timeline",
    "meta": Object {
      "cursor": "NA",
      "issueId": "MDU6SXNzdWU0MDA1MTYyOTE=",
      "issueNumber": 20,
      "projects": Array [
        "web3studio-internal",
      ],
      "repo": "ConsenSys/web3studio-website",
      "type": "IssueComment",
      "user": Object {
        "id": "MDQ6VXNlcjE2MjQ=",
        "login": "breakpointer",
      },
    },
    "timestamp": "2019-04-18T17:20:16Z",
  },
]
`;
